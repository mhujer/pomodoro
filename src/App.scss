html {
    // from https://www.typescriptlang.org/
    --code-font: 'Cascadia Mono-SemiLight', Menlo, Monaco, Consolas, monospace;
    --body-font: 'Segoe UI Web (West European)', 'Segoe UI', -apple-system, BlinkMacSystemFont, Roboto, 'Helvetica Neue',
        sans-serif;
    --text-color: #333;
    --background-color: #faf9f8;

    background-color: var(--background-color);
    color: var(--text-color);
    font-family: var(--body-font);
    font-size: 16px;
    line-height: 1.4;
}

body {
    display: flex;
    align-items: center;
    justify-content: center;
    flex-direction: column;
}

.container {
    text-align: center;
}

.countdown {
    font-weight: bold;
    font-size: 300%;
}

button {
    font-size: 170%;
    font-weight: bold;
    padding: 1em;
    box-shadow: 3px 3px 3px 0px rgba(0, 0, 0, 0.75);
}

footer {
    font-size: 85%;
    margin-top: 4em;
}
