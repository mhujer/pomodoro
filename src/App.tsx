import { useEffect, useRef, useState } from 'react';
import './App.scss';
import ding from '../src/sounds/ding.mp3'; // https://freesound.org/people/domrodrig/sounds/116779/
import Timer from './Timer/Timer';
import { formatPomodoroTime } from './timeFormatUtils';
import { PastPomodoro, PomodoroTimer } from './PomodoroTimer';
import { PastPomodoros } from './PastPomodoros/PastPomodoros';
import StartStopButton from './StartStopButton/StartStopButton';

const defaultTitle = document.title;

function App() {
    // @ts-expect-error startTime state is required to trigger rerender
    // eslint-disable-next-line @typescript-eslint/no-unused-vars
    const [startTime, setStartTime] = useState<number | null>(null);
    // @ts-expect-error now state is required to trigger rerender
    // eslint-disable-next-line @typescript-eslint/no-unused-vars
    const [now, setNow] = useState<number | null>(null);
    const [pastPomodoros, setPastPomodoros] = useState<PastPomodoro[]>([]);
    const intervalRef = useRef<number | null>(null);

    const pomodoroTimerRef = useRef<PomodoroTimer>(new PomodoroTimer());
    const pomodoroTimer = pomodoroTimerRef.current;

    function handleStart(): void {
        pomodoroTimer.startTimer();

        setStartTime(pomodoroTimer.startTime);
        setNow(pomodoroTimer.latestNow);

        if (intervalRef.current !== null) {
            clearInterval(intervalRef.current);
        }
        // eslint-disable-next-line @typescript-eslint/no-misused-promises
        intervalRef.current = window.setInterval(async () => {
            pomodoroTimer.tick();

            setStartTime(pomodoroTimer.startTime);
            setNow(pomodoroTimer.latestNow); // to trigger rerender

            // timer expired in tick()
            if (pomodoroTimer.startTime === null) {
                if (intervalRef.current !== null) { // guarded by TS
                    clearInterval(intervalRef.current);
                    intervalRef.current = null;
                }

                const myAudio = new Audio(ding);
                await myAudio.play();

                setPastPomodoros(pomodoroTimer.pastPomodoros);
            }
        }, 100);
    }

    function handleStop(): void {
        pomodoroTimer.stopTimer();

        if (intervalRef.current !== null) {
            clearInterval(intervalRef.current);
        }

        setStartTime(pomodoroTimer.startTime);
        setNow(pomodoroTimer.latestNow);
        setPastPomodoros(pomodoroTimer.pastPomodoros);
    }

    let remainingSeconds = pomodoroTimer.getRemainingSeconds();
    let title: string;
    if (remainingSeconds !== null) {
        title = `${formatPomodoroTime(remainingSeconds)} ${defaultTitle}`;
    } else {
        remainingSeconds = pomodoroTimer.getDefaultPomodoroDurationSeconds();
        title = defaultTitle;
    }

    // update title
    useEffect(() => {
        document.title = title;
    }, [title]);

    return (
        <div className='container'>
            <Timer remainingSeconds={remainingSeconds} />

            <StartStopButton
                isPomodoroActive={pomodoroTimer.isPomodoroActive()}
                handleStart={handleStart}
                handleStop={handleStop}
            />

            <PastPomodoros pastPomodoros={pastPomodoros} />
        </div>
    );
}

export default App;
