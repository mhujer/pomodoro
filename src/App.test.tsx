import React from 'react';
import App from './App';
import renderer from 'react-test-renderer';

it('renders App', () => {
    const domTree = renderer.create(<App />).toJSON();
    expect(domTree).toMatchSnapshot();
});
