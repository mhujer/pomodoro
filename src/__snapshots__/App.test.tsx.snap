// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders App 1`] = `
<div
  className="container"
>
  <p
    className="countdown"
  >
    00:05
  </p>
  <button
    onClick={[Function]}
  >
    Start!
  </button>
</div>
`;
